drivers/gpu/drm/rp1/rp1-dpi/rp1_dpi_pio.c

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * PIO code for Raspberry Pi RP1 DPI driver
+ *
+ * Copyright (c) 2024 Raspberry Pi Limited.
+ */
+
+/*
+ * Use PIO to fix up VSYNC for interlaced modes.
+ *
+ * For this to work we *require* DPI's pinctrl to enable DE on GPIO1.
+ * PIO can then snoop on HSYNC and DE pins to generate corrected VSYNC.
+ *
+ * Note that corrected VSYNC outputs will not be synchronous to DPICLK,
+ * will lag HSYNC by about 30ns and may suffer up to 5ns of jitter.
+ */
+
+#include <linux/kernel.h>
+#include <linux/errno.h>
+#include <linux/of.h>
+#include <linux/pio_rp1.h>
+#include <linux/pinctrl/consumer.h>
+#include <linux/platform_device.h>
+#include <drm/drm_print.h>
+
+#include "rp1_dpi.h"
+
+/*
+ * Start a PIO SM to generate an interrupt just after HSYNC onset, then another
+ * after a fixed delay (during which we assume HSYNC will have been deasserted).
+ */
+
+static int rp1dpi_pio_start_timer_both(struct rp1_dpi *dpi, u32 flags, u32 tc)
+{
+	static const u16 instructions[2][5] = {
+		{ 0xa022, 0x2083, 0xc001, 0x0043, 0xc001 }, /* posedge */
+		{ 0xa022, 0x2003, 0xc001, 0x0043, 0xc001 }, /* negedge */
+	};
+	const struct pio_program prog = {
+		.instructions = instructions[(flags & DRM_MODE_FLAG_NHSYNC) ? 1 : 0],
+		.length = ARRAY_SIZE(instructions[0]),
+		.origin = -1
+	};
+	int offset, sm;
+
+	sm = pio_claim_unused_sm(dpi->pio, true);
+	if (sm < 0)
+		return -EBUSY;
+
+	offset = pio_add_program(dpi->pio, &prog);
+	if (offset == PIO_ORIGIN_ANY)
+		return -EBUSY;
+
+	pio_sm_config cfg = pio_get_default_sm_config();
+
+	pio_sm_set_enabled(dpi->pio, sm, false);
+	sm_config_set_wrap(&cfg, offset, offset + 4);
+	pio_sm_init(dpi->pio, sm, offset, &cfg);
+
+	pio_sm_put(dpi->pio, sm, tc - 4);
+	pio_sm_exec(dpi->pio, sm, pio_encode_pull(false, false));
+	pio_sm_exec(dpi->pio, sm, pio_encode_out(pio_y, 32));
+	pio_sm_set_enabled(dpi->pio, sm, true);
+
+	return 0;
+}
+
+/*
+ * Snoop on DE, HSYNC to count half-lines in the vertical blanking interval
+ * to determine when the VSYNC pulse should start and finish. Then, at a
+ * suitable moment (which should be an odd number of half-lines since the
+ * last active line), sample DE again to detect field phase.
+ *
+ * This version assumes VFP length is within 2..129 half-lines for any field
+ * (one half-line delay is needed to sample DE; we always wait for the next
+ * half-line boundary to improve VSync start accuracy).
+ */
+
+static int rp1dpi_pio_vsync_ilace(struct rp1_dpi *dpi,
+				  struct drm_display_mode const *mode)
+{
+	static const int wrap_target = 14;
+	static const int wrap = 26;
+	u16 instructions[] = {  /* This is mutable */
```

Comment:
Using statics means we can't instantiate the driver twice. Unlikely ever to happen, but I'd be tempted to put it in struct rpi1_dpi instead.

Reply:
These are just named constants. The PIO compiler spits out `#define`s but I avoided those because they seemed unsightly, and there were formerly 3 different functions that defined them differently.

Reply:
I'd misread `instructions` as being static.

I assume the program gets copied into RP1 / PIO in some shape or form, so this doesn't need to remain in context after the `pio_add_program` call.

Reply:
Ah, it's the other way around: instructions[] is mutable because it gets subjected to some runtime fixups (rather than have 4 or 8 different versions for different sync polarities). But I can change the style if it's unclear.